File: server/src/main/java/org/tctalent/server/service/db/LinkPreviewService.java
/*
 * Copyright (c) 2024 Talent Catalog.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.service.db;

import org.tctalent.server.model.db.LinkPreview;

/**
 * Implements Jsoup web-scraping and HTML parsing to return link previews.
 * Uses <a href="https://andrejgajdos.com/how-to-create-a-link-preview/">...</a> as a starting point
 * for selecting content from the many divergent options available across different websites.
 *
 * @author samschlicht
 */

public interface LinkPreviewService {

  /**
   * Receives a URL in String form and returns a LinkPreview for display in Job Chats if successful.
   * @param url URL String of link to be previewed
   * @return Map of LinkPreview or null if unsuccessful
   */
  LinkPreview buildLinkPreview(String url);

  /**
   * @param linkPreviewId ID of record to be deleted
   * @return boolean - true if successful, false if not
   */
  boolean deleteLinkPreview(long linkPreviewId);

}

File: ui/admin-portal/src/app/services/link-preview.service.ts
/*
 * Copyright (c) 2024 Talent Catalog.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {environment} from '../../environments/environment';
import {HttpClient} from '@angular/common/http';
import {LinkPreview} from "../model/link-preview";

export interface BuildLinkPreviewRequest {
  url: string;
}

@Injectable({providedIn: 'root'})
export class LinkPreviewService {

  private apiUrl = environment.apiUrl + '/link-preview';

  constructor(private http: HttpClient) {}

  delete(id: number): Observable<boolean>  {
    return this.http.delete<boolean>(`${this.apiUrl}/${id}`);
  }

  buildLinkPreview(request: BuildLinkPreviewRequest): Observable<LinkPreview> {
    return this.http.post<LinkPreview>(`${this.apiUrl}/build-link-preview`, request)
  }
}
